content/technotes/2020-04-30-code-highlighting/index.Rmd

```diff
@@ -0,0 +1,411 @@
+
+```r {hl_lines=[1,"4-5"],linenos=table,linenostart=3}
+library("dplyr")
+df %>%
+  mutate(date = lubridate::ymd(date_string)) %>%
+  select(- date_string)
+str(df)
+nrow(df)
+```
+
+You can also configure line numbering [for your whole website](https://gohugo.io/getting-started/configuration-markup#highlight).
+
+The real magic to me is that if you write your code from R Markdown you can
+
+* apply the options to the source chunk using a knitr hook like the one defined [in our archetype](https://github.com/ropensci/roweb2/blob/master/archetypes/Rmd/index.md);
+
+* use R code to programmatically produce code block between fences, e.g. choosing which lines to highlight.
+
+#### knitr hook to highlight lines of source code
+
+Our hook is
+
+```r
+# knitr hook to use Hugo highlighting options
+knitr::knit_hooks$set(
+  source = function(x, options) {
+  hlopts <- options$hlopts
+    paste0(
+      "```", "r ",
+      if (!is.null(hlopts)) {
+      paste0("{",
+        glue::glue_collapse(
+          glue::glue('{names(hlopts)}={hlopts}'),
+          sep = ","
+        ), "}"
+        )
+      },
+      "\n", glue::glue_collapse(x, sep = "\n"), "\n```\n"
+    )
+  }
+)
+```
+
+The chunk[^chunk]
+
+````markdown
+```{r name-your-chunks, hlopts=list(linenos="table")} `r ''`
+a <- 1+1
+b <- 1+2
+c <- 1+3
+a + b + c
+```
+````
+
+is rendered as
+
+```{r name-your-chunks, hlopts=list(linenos="table")}
+a <- 1+1
+b <- 1+2
+c <- 1+3
+a + b + c
+```
+
+> PSA! Note that if you're after line highlighting, or function highlighting, for R Markdown documents _in general_, you should check out *[Kelly Bodwin's flair package](https://kbodwin.github.io/flair/index.html)*!
+
+#### Produce line-highlighted code blocks with glue/paste0
+
+What Chroma highlights are code blocks with code fences, which you might as well generate from R Markdown using some string manipulation and knitr `results="asis"` chunk option. E.g.
+
+````markdown
+
+```{r, results="asis"} `r ''`
+script <- c(
+  "a <- 1",
+  "b <- 2",
+  "c <- 3",
+  "a + b + c")
+cool_lines <- sample(1:4, 2)
+cool_lines <- stringr::str_remove(toString(cool_lines), " ")
+fences_start <- paste0('```', 'r {hl_lines=[', cool_lines,']}')
+glue::glue_collapse(
+  c(fences_start,  script,  "```"),
+sep = "\n")
+```
+
+````
+
+will be knit to produce
+
+```{r, results="asis", echo=FALSE}
+script <- c(
+  "a <- 1",
+  "b <- 2",
+  "c <- 3",
+  "a + b + c")
+cool_lines <- sample(1:4, 2)
+cool_lines <- stringr::str_remove(toString(cool_lines), " ")
+fences_start <- paste0('```', 'r {hl_lines=[', cool_lines,']}')
+glue::glue_collapse(
+  c(fences_start,  script,  "```"),
+sep = "\n")
+```
+
+This is a rather uninteresting toy example since we used randomly drawn line numbers to be highlighted, but you might find use cases for this.
+We used such an approach in the recent [blog post about Rclean](/blog/2020/04/21/rclean/), actually!
+
+#### Accessible emphasis
```

Comment:
I would move all discussion about Accessibility to a section here were you talk about highlighting and line numbering together. 

Also, what about making sure to keep commenting code blocks for people who are using screen readers?

Reply:
> Also, what about making sure to keep commenting code blocks for people who are using screen readers?

Do you mean explaining them? Don't screen readers read the code blocks? should I link to https://r-resources.massey.ac.nz/rmarkdown/?

Reply:
Sorry, I meant that you say people should still comment their code blocks in case the highlights aren't obvious, and I meant that you could also add, that for people with screen readers, the highlights won't be seen at all, so another good reason to comment code blocks.

Reply:
And that link would be great I think! Even if just to alert people to think about it, without having to get into the nitty gritty.